src/ValidatorKeys.cpp

```diff
@@ -214,20 +259,58 @@ ValidatorKeys::revoke ()
     st[sfSequence] = std::numeric_limits<std::uint32_t>::max ();
     st[sfPublicKey] = publicKey_;
 
-    ripple::sign(st, HashPrefix::manifest, keyType_, secretKey_,
-        sfMasterSignature);
+    ripple::sign(st, HashPrefix::manifest, keyType_, secretKey_, sfMasterSignature);
 
     Serializer s;
     st.add(s);
 
-    std::string m (static_cast<char const*> (s.data()), s.size());
-    return beast::detail::base64_encode(m);
+    manifest_.clear();
+    manifest_.reserve(s.size());
+    std::copy(s.begin(), s.end(), std::back_inserter(manifest_));
+
+    return beast::detail::base64_encode(manifest_.data(), manifest_.size());
 }
 
 std::string
-ValidatorKeys::sign (std::string const& data)
+ValidatorKeys::sign (std::string const& data) const
 {
     return strHex(ripple::sign (publicKey_, secretKey_, makeSlice (data)));
 }
 
+void
+ValidatorKeys::domain(std::string d)
+{
+    if (!d.empty())
+    {
+        // A valid domain for a validator must be at least 4 characters
+        // long, should contain at least one . and should not be longer
+        // that 128 characters.
+        if (boost::algorithm::clamp(d.size(), 4, 128) != d.size())
+            throw std::runtime_error (
+                "The domain must be between 4 and 128 characters long.");
+
+        // This regular expression should do a decent job of weeding out
+        // obviously wrong domain names but it isn't perfect. It does not
+        // really support IDNs. If this turns out to be an issue, a more
+        // thorough regex can be used or this check can just be removed.
+        static boost::regex const re(
+                "^"                     // Beginning of line
+                "("                     // Hostname or domain name
+                "(?!-)"                 //  - must not begin with '-'
+                "[a-zA-Z0-9-]{1,63}"    //  - only alphanumeric and '-'
+                "(?<!-)"                //  - must not end with '-'
+                "\\."                   // segment separator
+                ")+"                    // 1 or more segments
+                "[A-Za-z]{2,63}"        // TLD
```

Comment:
hmm - TLDs can be punycode, right? Do we care about ascii-only for this field? We're not going to try to resolve it, are we? Alternatively, we might consider just IDNA2008 encoding anything high-bit. I added a testcase that seems like it should work, but does not currently:  https://github.com/mellery451/validator-keys-tool/commit/e1917fc3a85cd0582fc4ef5796e5fc0fd0b6352b